### ISSUE TITLE
[❔ other question] Specify Vector3f in put_parameter

### ISSUE THREAD

**Issue by JulianKnodt** [mitsuba-renderer/mitsuba2#248]
## Summary

I was wondering if there was a way to add a differentiable `Vector3f` in `put_parameter`. When I add it to the list of parameters, it appears that it is non-differentiable. Is that because of how I use it, or is it due to its type?

**Comment by Speierers** [comment]
Regarding the `Vector3f`, this might be a bug, although the variable you would access in this case should still be differentiable.

For a variable to show up in `graphviz`, you will need to mark it as requiring gradients.

There should be a way to call `graphviz` from cpp: https://github.com/mitsuba-renderer/enoki/blob/master/include/enoki/autodiff.h#L103

Not sure what you are trying to achieve with `DynamicBuffer<ScalarVector3f>`? Isn't `Vector3f` already a dynamic type here?

**Comment by JulianKnodt** [comment]
Er, what do you mean that it should still be differentiable? Even if it's not indicated with an `*` in the list of parameters in python, I can still compute gradients with respect to it and backpropagate? I'm not sure if when I was testing it out if it showed up at all in the list or if it caused an error, I'll need to test again.

I think even after calling `set_require_gradient` on some variables, I found that they could not be in

Ah, I've been using `mesh.h/mesh.cpp` & `shapes/ply.cpp` as reference for how they use a `DynamicBuffer<replace_scalar_t<Float, float>>`. What I'm trying to do is store an array of vectors in a DynamicBuffer, and then perform gradient descent w.r.t. the parameters in the vector. Even though I currently know the number of items in the array, having it be dynamic is nice as I might want to change it in the future. I know the structure of my program is fine, as if I pass a mask of `false` to `gather`, it runs properly, but once I pass `true`, it segfaults inside of gather.

**Comment by Speierers** [comment]
First of all, `replace_scalar_t<Float, float>` will just return type `Float` as the internal `Scalar` type of `Float` is `float` already.
Enoki mainly works with Structure of Array (SoA), so what you actually want is a vector of dynamic arrays (and not a dynamic array of vectors). And this is exactly what `Vector3f` (a.k.a. `Array<DiffArray<:CUDAArray<float>>, 3>`) is. 
I guess your call to `gather` is failing as it is expecting a `SoA` input.

**Comment by JulianKnodt** [comment]
Hm, ok, how would I go about doing that?
I've been looking [here](https://github.com/mitsuba-renderer/mitsuba2/blob/74253f4edd8a3cbc8711ad31abd65fc6d7dcb81d/include/mitsuba/render/mesh.h#L94), but is there some other location to use as reference? I've also been looking at the docs, but there's not many generalizable examples.

**Comment by Speierers** [comment]
`vertex_normals()` gathers `Normal3f` from a flat array of float which is different. Where do you need to gather your vectors from?

**Comment by JulianKnodt** [comment]
I'm fine with anything differentiable, so I can just use a flat array of floats as well.

**Comment by Speierers** [comment]
> What I also don't understand is that the call to gather actually succeeds once, but fails on the second iteration? I imagine if my invokation was incorrect it would fail both times.

That I won't be able to tell for sure. It wasn't design for this, but it might well be that it still works in some cases.

Should I close this issue?

**Comment by JulianKnodt** [comment]
I mean I guess it's fine, my issue isn't really resolved but I guess I'm trying to use it in a way it wasn't intended for.

**Comment by wjakob** [comment]
There are various plugins that add differentiable 3D vectors (check e.g. src/spectra/srgb.cpp) -- can you explain how what you're doing is different?

I believe you may need a cuda_sync() after the `.managed()` call.

**Comment by JulianKnodt** [comment]
Hm, well I'm also using other sized vectors. I realize that it's only the case when I'm using Vectors with length greater than 3 that it fails

**Comment by JulianKnodt** [comment]
Let me see if adding a `cuda_sync` after managed makes it work
